libyul/optimiser/StackCompressor.cpp

```diff
@@ -0,0 +1,93 @@
+/*(
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimisation stage that aggressively rematerializes certain variables ina a function to free
+ * space on the stack until it is compilable.
+ */
+
+#include <libyul/optimiser/StackCompressor.h>
+
+#include <libyul/optimiser/SSAValueTracker.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/Rematerialiser.h>
+#include <libyul/optimiser/UnusedPruner.h>
+#include <libyul/optimiser/Metrics.h>
+#include <libyul/optimiser/Semantics.h>
+
+#include <libyul/CompilabilityChecker.h>
+
+#include <libyul/AsmData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+bool StackCompressor::run(Block& _ast)
+{
+	for (size_t iterations = 0; iterations < 4; iterations++)
+	{
+		map<YulString, int> stackSurplus = CompilabilityChecker::run(m_dialect, _ast);
+		if (stackSurplus.empty())
+			return true;
+
+		if (stackSurplus.count(YulString{}))
+		{
+			// TODO Main block is not compilable - this is not relevant for now.
```

Comment:
This actually means that it will just not run on most tests (they usually use a big main part because all functions would be inlined anyway and we cannot force functions to not be touched yet), so probably better to implement it.

Idea for implementation: Either visit individual non-functiondefinition statements or instruct analysis tools below to not go inside function definitions.

Either way it's not that nice.

We could also extract all non-functiondefinition statements into a new block (or even function) and run it there.

Reply:
Fixed this.